ledger/acctonline.go

```diff
@@ -0,0 +1,846 @@
+	// the invariant is that the last element of ao.onlineRoundParamsData is for round ao.latest()
+	startRound := ao.latest() + 1 - basics.Round(len(ao.onlineRoundParamsData))
+	if rnd < startRound {
+		err = &RoundOffsetError{
+			round:   rnd,
+			dbRound: startRound,
+		}
+		return
+	}
+
+	off := uint64(rnd - startRound)
+	if off >= uint64(len(ao.onlineRoundParamsData)) {
+		err = fmt.Errorf("round %d too high: dbRound %d, onlineRoundParamsData %d", rnd, startRound, len(ao.onlineRoundParamsData))
+		return
+	}
+
+	return off, nil
+}
+
+// lookupOnlineAccountData returns the online account data for a given address at a given round.
+func (ao *onlineAccounts) lookupOnlineAccountData(rnd basics.Round, addr basics.Address) (data ledgercore.OnlineAccountData, err error) {
+	ao.accountsMu.RLock()
+	needUnlock := true
+	defer func() {
+		if needUnlock {
+			ao.accountsMu.RUnlock()
+		}
+	}()
+	var offset uint64
+	var paramsOffset uint64
+	var rewardsProto config.ConsensusParams
+	var rewardsLevel uint64
+	var persistedData persistedOnlineAccountData
+
+	for {
+		currentDbRound := ao.cachedDBRoundOnline
+		currentDeltaLen := len(ao.deltas)
+		offset, err = ao.roundOffset(rnd)
+		inHistory := false
+		if err != nil {
+			var roundOffsetError *RoundOffsetError
+			if !errors.As(err, &roundOffsetError) {
+				return
+			}
+			// the round number cannot be found in deltas, it is in history
+			inHistory = true
+			err = nil
+		}
+		paramsOffset, err = ao.roundParamsOffset(rnd)
+		if err != nil {
+			return
+		}
+
+		rewardsProto = config.Consensus[ao.onlineRoundParamsData[paramsOffset].CurrentProtocol]
+		rewardsLevel = ao.onlineRoundParamsData[paramsOffset].RewardsLevel
+
+		// check if we've had this address modified in the past rounds. ( i.e. if it's in the deltas )
+		if !inHistory {
+			macct, indeltas := ao.accounts[addr]
+			if indeltas {
+				// Check if this is the most recent round, in which case, we can
+				// use a cache of the most recent account state.
+				if offset == uint64(len(ao.deltas)) {
+					return macct.data.OnlineAccountData(rewardsProto, rewardsLevel), nil
+				}
+				// the account appears in the deltas, but we don't know if it appears in the
+				// delta range of [0..offset], so we'll need to check :
+				// Traverse the deltas backwards to ensure that later updates take
+				// priority if present.
+				for offset > 0 {
+					offset--
+					d, ok := ao.deltas[offset].GetData(addr)
+					if ok {
+						return d.OnlineAccountData(rewardsProto, rewardsLevel), nil
+					}
+				}
+			}
+		}
+
+		if macct, has := ao.onlineAccountsCache.read(addr, rnd); has {
+			return macct.GetOnlineAccountData(rewardsProto, rewardsLevel), nil
+		}
+
+		ao.accountsMu.RUnlock()
+		needUnlock = false
+
+		// No updates of this account in the in-memory deltas; use on-disk DB.
+		// As an optimization, we avoid creating
+		// a separate transaction here, and directly use a prepared SQL query
+		// against the database.
```

Comment:
Move this comment to `lookupOnline`, if it's important to indicate that it's a prepared query you could rename `accountsPreparedQueries`